todo/services/role_service.py

```diff
@@ -0,0 +1,102 @@
+from todo.exceptions.role_exceptions import (
+    RoleNotFoundException,
+    RoleAlreadyExistsException,
+    RoleOperationException,
+)
+
+
+class RoleService:
+    @classmethod
+    def get_all_roles(cls, filters: Optional[Dict[str, Any]] = None) -> List[RoleDTO]:
+        try:
+            role_models = RoleRepository.list_all(filters=filters)
+            return [RoleDTO.from_model(role) for role in role_models]
+        except Exception as e:
+            raise RoleOperationException(f"Failed to get roles: {str(e)}")
+
+    @classmethod
+    def get_role_by_id(cls, role_id: str) -> RoleDTO:
+        role_model = RoleRepository.get_by_id(role_id)
+        if not role_model:
+            raise RoleNotFoundException(role_id)
+        return RoleDTO.from_model(role_model)
+
+    @classmethod
+    def create_role(
+        cls,
+        name: str,
+        description: Optional[str],
+        role_type: str,
+        scope: str,
+        is_active: bool,
+        created_by: str,
+    ) -> RoleDTO:
+        try:
+            role_model = RoleModel(
+                name=name,
+                description=description,
+                type=RoleType(role_type),
+                scope=RoleScope(scope),
+                is_active=is_active,
+                created_by=created_by,
+                created_at=datetime.now(timezone.utc),
+            )
+
+            created_role = RoleRepository.create(role_model)
+            return RoleDTO.from_model(created_role)
+
+        except ValueError as e:
+            role_name = str(e).split("'")[1] if "'" in str(e) else "unknown role"
+            raise RoleAlreadyExistsException(role_name)
+        except Exception as e:
+            raise RoleOperationException(f"Failed to create role: {str(e)}")
+
+    @classmethod
+    def update_role(cls, role_id: str, **update_data) -> RoleDTO:
+        try:
+            existing_role = RoleRepository.get_by_id(role_id)
+            if not existing_role:
+                raise RoleNotFoundException(role_id)
+
+            clean_update_data = {k: v for k, v in update_data.items() if v is not None}
+
+            if "type" in clean_update_data and isinstance(clean_update_data["type"], str):
+                clean_update_data["type"] = RoleType(clean_update_data["type"])
+            if "scope" in clean_update_data and isinstance(clean_update_data["scope"], str):
+                clean_update_data["scope"] = RoleScope(clean_update_data["scope"])
+
+            updated_role = RoleRepository.update(role_id, clean_update_data)
+            if not updated_role:
+                raise RoleOperationException(f"Failed to update role with ID: {role_id}")
+
+            return RoleDTO.from_model(updated_role)
+
+        except ValueError as e:
+            raise RoleAlreadyExistsException(str(e))
+        except RoleNotFoundException:
+            raise
+        except Exception as e:
+            raise RoleOperationException(f"Failed to update role: {str(e)}")
```

Comment:
- if we have a global exception handler, can we define it there?
for example: https://github.com/Real-Dev-Squad/skill-tree-backend/blob/develop/skill-tree/src/main/java/com/RDS/skilltree/exceptions/GlobalExceptionHandler.java

Reply:
done